tiny_dnn/core/kernels/global_avepool_op_avx.h

```diff
@@ -0,0 +1,120 @@
+    const vec_t &in = in_data[sample];
+    vec_t &out      = out_data[sample];
+
+    for (size_t i = 0; i < params.in.depth_; i++) {
+      __m256 sum_m             = _mm256_setzero_ps();
+      const size_t depth_index = i * pool_area;
+      for (size_t j = 0; j < nblocks_per_channel; j++) {
+        __m256 in_m = _mm256_load_ps(&in[depth_index + 8 * j]);
+        sum_m       = _mm256_add_ps(sum_m, in_m);
+      }
+      if (nremains_per_channel) {
+        __m256 in_m = _mm256_setzero_ps();
+        in_m =
+          _mm256_maskload_ps(&in[depth_index + 8 * nblocks_per_channel], imask);
+        sum_m = _mm256_add_ps(sum_m, in_m);
+      }
+      out[i] = _mm_cvtss_f32(hsum256_ps(sum_m));
+      out[i] *= pool_area_inv;
+    }
+  });
+}
+
+inline void avx_global_avepool_backward_kernel(
+  tensor_t &prev_delta,
+  const tensor_t &curr_delta,
+  const core::global_avepool_params &params,
+  const bool layer_parallelize) {
+  const size_t pool_area           = params.in.width_ * params.in.height_;
+  const size_t pool_area_inv       = 1.0f / pool_area;
+  const size_t nblocks_per_channel = pool_area / 8;
+
+  for_i(layer_parallelize, prev_delta.size(), [&](int sample) {
+    vec_t &prev       = prev_delta[sample];
+    const vec_t &curr = curr_delta[sample];
+
+    for (size_t i = 0; i < params.in.depth_; i++) {
+      const float_t pi = curr[i] * pool_area_inv;
+      size_t j         = 0;
+      __m256 prev0     = _mm256_set1_ps(pi);
+
+      while (j < nblocks_per_channel) {
+        _mm256_store_ps(&prev[8 * j], prev0);
+        j++;
+      }
+
+      for (size_t k = 8 * j; k < pool_area; k++) {
+        prev[i * pool_area + k] = pi;
+      }
+    }
+  });
+}
+
+#endif  // CNN_USE_AVX
+
+inline void global_avepool_op_avx(const tensor_t &in_data,
+                                  tensor_t &out_data,
+                                  const core::global_avepool_params &params,
+                                  const bool layer_parallelize) {
+#ifdef CNN_USE_AVX
+  avx_global_avepool_forward_kernel(in_data, out_data, params,
+                                    layer_parallelize);
+#else
+  CNN_UNREFERENCED_PARAMETER(in_data);
+  CNN_UNREFERENCED_PARAMETER(out_data);
+  CNN_UNREFERENCED_PARAMETER(params);
+  CNN_UNREFERENCED_PARAMETER(layer_parallelize);
+  throw nn_error("TinyDNN has not been compiled with AVX support.");
+#endif
+}
+
+inline void global_avepool_grad_op_avx(
+  tensor_t &prev_delta,
+  const tensor_t &curr_delta,
+  const core::global_avepool_params &params,
+  const bool layer_parallelize) {
+#ifdef CNN_USE_AVX
+  avx_global_avepool_backward_kernel(prev_delta, curr_delta, params,
+                                     layer_parallelize);
+#else
+  CNN_UNREFERENCED_PARAMETER(prev_delta);
+  CNN_UNREFERENCED_PARAMETER(curr_delta);
+  CNN_UNREFERENCED_PARAMETER(params);
+  CNN_UNREFERENCED_PARAMETER(layer_parallelize);
+  throw nn_error("TinyDNN has not been compiled with AVX support.");
```

Comment:
I'd write fallback call to internal backend routine `global_avepool_grad_op_internal` instead of throwing an exceotion.

In that way, you can remove backend type check in `global_average_pooling_layer::init_backend` method in `tiny_dnn/layers/global_average_pooling_layer.h` file.